### PR TITLE
Allow null group assignment for users (Issue #867)

### PR DESCRIPTION
**Pull request by amosfolz**
I think this is ready for a review. 
Maybe someone else can suggest a better way to handle showing an empty option using select2. 

Ref.: https://github.com/userfrosting/UserFrosting/issues/867

### CODE REVIEW

**Comment by lcharette** [comment]
Looks like something is broken with the new migration : https://travis-ci.org/userfrosting/UserFrosting/jobs/518564703#L873

**Comment by amosfolz** [comment]
I am going to try to see if I can figure out what is causing Travis to fail here. I noticed the migrations are all completing without any errors...
image

**Comment by lcharette** [comment]
The good news is, you're not crazy. I checked out your branch and can't replicate the Travis issue. Even if I replicate Travis env locally (ie not using the `:memory:`database), the only issue I see is related to `testGetAsset`.

**Comment by lcharette** [comment]
One thing to note, Travis complains about [`UpdateUsersTable.php:56`](https://travis-ci.org/userfrosting/UserFrosting/jobs/519717573#L967), which means the issue would be with the `down` method...

Actually, it could be a MySQL config thing. [The error is different on PgSQL](https://travis-ci.org/userfrosting/UserFrosting/jobs/519717583#L918). Something might be different between Travis MySQL setup and local/Homestead one.

**Comment by amosfolz** [comment]
Okay, so I believe the PgSQL error is actually related to this https://github.com/userfrosting/assets/releases/tag/5.0.4

**Comment by lcharette** [comment]
> Okay, so I believe the PgSQL error is actually related to this https://github.com/userfrosting/assets/releases/tag/5.0.4

And I believe it was fixed in `hotfix` here : https://github.com/userfrosting/UserFrosting/commit/36ed889c7ef197cf060f5a2d4f69e523aada94e8

Since this PR target UF 4.3, I suggest we leave it as is for now and look at it once we have released 4.2.1. We'll need to merge hotfix into develop at some point anyway.

**Comment by amosfolz** [comment]
Is this something in the migration that needs to be fixed or is there possibly something in the test environment/configuration causing this?

**Comment by amosfolz** [comment]
I think this will need to be a non-reversible change. I have been trying for hours to get a down migration to work that will set the column back to `not null` and have had no success.

**Comment by lcharette** [comment]
> I think this will need to be a non-reversible change. I have been trying for hours to get a down migration to work that will set the column back to `not null` and have had no success.

If the change in the db is not reversed, and the files are reversed to 4.2.x, does it still works? Obviously, you won't be able to select "no group", but as long as the db still works, we should be good.

**Comment by amosfolz** [comment]
> 
> 
> > I think this will need to be a non-reversible change. I have been trying for hours to get a down migration to work that will set the column back to `not null` and have had no success.
> 
> If the change in the db is not reversed, and the files are reversed to 4.2.x, does it still works? Obviously, you won't be able to select "no group", but as long as the db still works, we should be good.

It should still work.

If that happened the only difference would be `group_id` would remain `nullable`. 

```
   $table->unsignedInteger('group_id')->default(1)->comment('The id of the user group.')->change();
```

**Comment by lcharette** [comment]
Only issue I see so far is in the user detail page : 

Capture d’écran, le 2019-07-10 à 20 52 24

The ` •` separating the username and group makes it weird now. Simple if in Twig should fix that.

File: app/sprinkles/admin/templates/forms/user.html.twig [hunk 1]
```diff
@@ -24,7 +24,8 @@
                         {% if 'group' in form.fields.disabled %}
                             <input type="text" class="form-control" name="theme" value="{{user.group.name}}" disabled>
                         {% else %}
-                            <select id="input-group" class="form-control js-select2" name="group_id">
+                            <select id="input-group" class="form-control js-select2-group" name="group_id">
+                                <option value="0" </option>
```
Comment (lcharette) [inline app/sprinkles/admin/templates/forms/user.html.twig]: Is this valid html? Shouldn't it be `<option value="0"></option>` ?

File: app/sprinkles/admin/templates/forms/user.html.twig [hunk 2]
```diff
@@ -24,7 +24,8 @@
                         {% if 'group' in form.fields.disabled %}
                             <input type="text" class="form-control" name="theme" value="{{user.group.name}}" disabled>
                         {% else %}
-                            <select id="input-group" class="form-control js-select2" name="group_id">
+                            <select id="input-group" class="form-control js-select2-group" name="group_id">
```
Comment (lcharette) [inline app/sprinkles/admin/templates/forms/user.html.twig]: `js-select2-group` is not defined in this PR, making the select not select2 styled. 

Plus, when applying select2, the "no group" styling is weird. Should probably says "no group" or something.

Capture d’écran, le 2019-07-10 à 21 26 15

**Comment by amosfolz** [comment]
> 
> 
> Only issue I see so far is in the user detail page :
> 
> Capture d’écran, le 2019-07-10 à 20 52 24
> 
> The ` •` separating the username and group makes it weird now. Simple if in Twig should fix that.

Do we want this to show "No group" or just remove if they have that option select?

File: app/sprinkles/admin/templates/forms/user.html.twig (hunk 1 above)
Comment (amosfolz) [reply app/sprinkles/admin/templates/forms/user.html.twig]: Fixed in b26b70512089f7f21cd3d09f609684edf95cf9c5

File: app/sprinkles/admin/templates/forms/user.html.twig (hunk 2 above)
Comment (amosfolz) [reply app/sprinkles/admin/templates/forms/user.html.twig]: Fixed in b26b70512089f7f21cd3d09f609684edf95cf9c5

**Comment by lcharette** [comment]
> Do we want this to show "No group" or just remove if they have that option select?

I think removing it by default makes more sense if someone doesn't want to use groups at all.

File: app/sprinkles/admin/templates/forms/user.html.twig (hunk 2 above)
Comment (lcharette) [reply app/sprinkles/admin/templates/forms/user.html.twig]: The `No group`  will be the same formatting as the others...  Does Select2 support having a `<i>` inside the `option` element?

**Comment by amosfolz** [comment]
> 
> 
> Only issue I see so far is in the user detail page :
> 
> Capture d’écran, le 2019-07-10 à 20 52 24
> 
> The ` •` separating the username and group makes it weird now. Simple if in Twig should fix that.

Fixed in 3467690d4e52b78b6c945d5dfc2c7b2f08a72677.


> 
> 
> > I think this will need to be a non-reversible change. I have been trying for hours to get a down migration to work that will set the column back to `not null` and have had no success.
> 
> If the change in the db is not reversed, and the files are reversed to 4.2.x, does it still works? Obviously, you won't be able to select "no group", but as long as the db still works, we should be good.

I retested and can confirm the DB still works fine in this situation.

**Comment by lcharette** [comment]
> I retested and can confirm the DB still works fine in this situation.

I did too last and it does works fine.

File: app/sprinkles/admin/templates/forms/user.html.twig (hunk 2 above)
Comment (amosfolz) [reply app/sprinkles/admin/templates/forms/user.html.twig]: Oops... It does not look like `<i>` is supported inside `option`

**Pull request merged**